8295486: Inconsistent constant field values observed during compilation

TobiHartmann opened the pull request:
We hit a "not monotonic" assert because the new type of a load from a stable final field is more narrow than the old type which contradicts the assumption that types should only go from TOP to BOTTOM during CCP:

old: `narrowoop: java/lang/Integer:BotPTR:exact *`
new: `narrowoop: java/lang/Integer java.lang.Integer {0x000000062c41e548} ...`

or 

old: `narrowoop: java/lang/Integer java.lang.Integer {0x000000062c41e538} ...`
new: `narrowoop: java/lang/Integer java.lang.Integer {0x000000062c41e548} ...`

The problem is that a stable field can be (re-)initialized during compilation and since the value is not cached, contradicting types can be observed. In `LoadNode::Value`, we re-read the field value each time: 

https://github.com/openjdk/jdk/blob/872384707e89d03ede655aad16f360dc94f10152/src/hotspot/share/opto/memnode.cpp#L1994-L1997

https://github.com/openjdk/jdk/blob/872384707e89d03ede655aad16f360dc94f10152/src/hotspot/share/opto/type.cpp#L332-L337

The same problem exists for loads from stable arrays:
https://github.com/openjdk/jdk/blob/872384707e89d03ede655aad16f360dc94f10152/src/hotspot/share/opto/memnode.cpp#L1923

Caching the field value is not feasible as it would require a cache per ciInstance for all the fields and per ciArray for all the elements. Alternatively, we could keep track of the lookup and only do it once but that would also be lots of additional complexity for a benign issue.

Instead, I propose to skip verification during CCP when folding loads from stable fields. Non-stable, constant fields are not affected as `null` is a valid value for them and they would already be folded before CCP.

Thanks,
Tobias

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8295486](https://bugs.openjdk.org/browse/JDK-8295486): Inconsistent constant field values observed during compilation


### Reviewers
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**) ⚠️ Review applies to [32e3287e](https://git.openjdk.org/jdk/pull/11861/files/32e3287e0beb7966c3551f107a4ec2c79b4db660)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Jatin Bhateja](https://openjdk.org/census#jbhateja) (@jatin-bhateja - **Reviewer**) ⚠️ Review applies to [048b273f](https://git.openjdk.org/jdk/pull/11861/files/048b273fe40ab6db767c4a326d46290378f524aa)
 * [Vladimir Ivanov](https://openjdk.org/census#vlivanov) (@iwanowww - **Reviewer**) ⚠️ Review applies to [23344db9](https://git.openjdk.org/jdk/pull/11861/files/23344db98a864232b771350efa4edce947f7de5b)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11861/head:pull/11861` \
`$ git checkout pull/11861`

Update a local copy of the PR: \
`$ git checkout pull/11861` \
`$ git pull https://git.openjdk.org/jdk pull/11861/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11861`

View PR using the GUI difftool: \
`$ git pr show -t 11861`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11861.diff">https://git.openjdk.org/jdk/pull/11861.diff</a>

</details>

TobiHartmann commented:
Thanks, Christian!

vnkozlov reviewed (commented):
@vlivanov Do we  have a mechanism to verify that `Stable` fields are initialized only once? Based on Tobias's test (which continuously modifies values) it does not work if it even exists. Then what is definition of `Stable` attribute then? We can't constant fold a field's value which can be changed (except from NULL to a value).

Tobias's comment in JBS about `field point to different DirectMethodHandles` concerns me.

For me the only valid solution is throw out compilation if it see change in stable filed's state. We have validation step in `ciEnv::register_method()`. May be create table of stable fields values we folded and verify during registration to make sure it did not change. It may not catch case in Tobias's test where the same values are rotated.

Note, **one** change from  null to not-null is valid case. May be new code in `PhaseCCP::verify_type()` should check for initial NULL value before skipping following checks.

jatin-bhateja reviewed (commented):
Nice test!

vnkozlov commented:
> The definition of @Stable is basically that we can constant-fold any value once it's initialized and (semantically) it does not make a difference which value we choose.

Okay. The spec is really bizarre for @Stable

> If we go with your proposed solution of a table of field values, we can as well use that one as cache instead of bailing out from compilation, right?

Right. I prefer caching because the value will be consistent at least during compilation and it may allow some optimizations to happen. Imaging you read same stable field on separate paths and then compare constants.

iwanowww commented:
I fully support caching of non-default values. (Not sure about default values yet: at least, it makes sense to cache them as well to make compilations reproducible.) 

As an idea, the table could be built on top of existing dependency mechanism, but `@Stable`-related dependencies have to be dropped when migrated to the nmethod.

iwanowww commented:
> @vlivanov Do we have a mechanism to verify that Stable fields are initialized only once?

As Tobias already pointed out, there's no such support in place yet. Only a long-standing RFE to add it:
  https://bugs.openjdk.org/browse/JDK-8024042

jatin-bhateja commented:
> > The definition of @stable is basically that we can constant-fold any value once it's initialized and (semantically) it does not make a difference which value we choose.
> 
> Okay. The spec is really bizarre for @stable
> 
> > If we go with your proposed solution of a table of field values, we can as well use that one as cache instead of bailing out from compilation, right?
> 
> Right. I prefer caching because the value will be consistent at least during compilation and it may allow some optimizations to happen. Imaging you read same stable field on separate paths and then compare constants.

If value keep changing during CCP then it will delay the convergence. Thus a convergence will be reached only for a stable value,  which is what caching will also achieve. Stable filed value should be added to method dependencies such that compilation becomes invalid if stable value changes later on.

iwanowww commented:
> Stable field value should be added to method dependencies such that compilation becomes invalid if stable value changes later on.

It's too much for @Stable. It's allowed to observe stale values when concurrent modifications happen. It would require dependency checking on every `@Stable` field or array element update which is infeasible for arrays (all array stores have to be intercepted).

On the other hand, caching a single value which is used across the whole compilation has only performance impact (whether non-null value is observed or not).

jatin-bhateja commented:
> > Stable field value should be added to method dependencies such that compilation becomes invalid if stable value changes later on.
> 
> It's too much for @stable. It's allowed to observe stale values when concurrent modifications happen. It would require dependency checking on every `@Stable` field or array element update which is infeasible for arrays (all array stores have to be intercepted).
> 
> On the other hand, caching a single value which is used across the whole compilation has only performance impact (whether non-null value is observed or not).

Agree, but there should be a way for method to deoptimize if stable field value changes past method registration. Only full proof solution is to enforce that stable field value does not change after single initialisation apart from null to non-null transitions.

TobiHartmann commented:
Vladimirs, Jatin, thanks for the reviews and discussion!

The question is what we want to achieve with this patch:
1) Prevent C2 from crashing / asserting when observing inconsistent field values
2) Prevent C2 from applying optimizations based on inconsistent field values
3) Prevent C2 from constant folding stale values

I think my proposed patch achieves 1) which is also the only issue we ever observed in real code: Racy initialization of method handles for indy string concat in core libraries code. The verification code proposed by ([JDK-8024042](https://bugs.openjdk.org/browse/JDK-8024042)) would catch this. Constant folding a stale value does not matter here, semantically, we can use any of the equivalent method handles that we observe. It just confuses CCP verification. Convergence of CCP is also not an issue because we don't update a node's type indefinitely but only if the type of a relevant input node changed.

Regarding 2): Couldn't any such wrong/undesired behaviour happen with execution in the interpreter as well? The only difference would be that optimized C2 compiled code would **always** behave that way. But according to the specification for `@Stable`, that's okay:

https://github.com/openjdk/jdk/blob/5c8c67c523279de728248f54382c40fb20d0ab63/src/java.base/share/classes/jdk/internal/vm/annotation/Stable.java#L74-L80

Also, doesn't the above specification mean that all non-synchronized initializations of stable fields can lead to undefined behavior?

> Imaging you read same stable field on separate paths and then compare constants.

That will also happen with interpreted code when the stable field is written between the two reads.

The only complete solution would be 3), which, as Vladimir I. already pointed out, does not seem feasible given that we would need to intercept all writes to stable fields and array elements.

I'm not against caching but I'm wondering how much sense it makes to apply an expensive and complex (partial) fix to C2 while C1 and the interpreter are still affected. Shouldn't the remaining issues be fixed in Java code if undefined/unexpected behavior is ever observed?

iwanowww commented:
IMO we should focus solely on C2 where aforementioned inconsistencies trigger assertion failures and (rarely) break compilation replay. (The rest is classified as user errors according to `@Stable` contract.) Caching a single value (either null or non-null) for the duration of the compilation addresses both problems.

iwanowww commented:
> but there should be a way for method to deoptimize if stable field value changes past method registration.

It is not required for proper `@Stable` support and, moreover, doesn't make much sense considering how the API is shaped.

`@Stable` is not part of public API (has no effect in non-trusted code) for a reason: it deliberately doesn't try to enforce the invariants it depends on and leaves users to struggle with the consequences without providing any help. So, caution is needed when it is used.

iwanowww commented:
FTR there have been multiple attempts to build a story for "lazy final" fields:
* https://bugs.openjdk.org/browse/JDK-8297156
* https://openjdk.org/jeps/8209964
* https://cr.openjdk.java.net/~jrose/draft/lazy-final.html

vnkozlov commented:
> IMO we should focus solely on C2 where aforementioned inconsistencies trigger assertion failures and (rarely) break compilation replay. (The rest is classified as user errors according to `@Stable` contract.) Caching a single value (either null or non-null) for the duration of the compilation addresses both problems.

+1

C2 generated code is "final" (C1 code and Interpreter are executed only at initial stage with exception of deoptimization).

TobiHartmann commented:
Fair enough. I implemented caching of stable values. Please let me know what you think.

vnkozlov reviewed (approved):
Nice work.

iwanowww commented on src/hotspot/share/ci/ciEnv.hpp:
```diff
@@ -466,6 +467,28 @@ class ciEnv : StackObj {
+      ciConstant value() const { return _value; }
+  };
+
+  GrowableArray<StableValue>* _stable_values; // Cache of stable values
```
Aren't `ciInstance` and `ciArray` provide a better place to cache field/element values (logic can be encapsulated in `ciInstance::field_value()` and `ciArray::element_value()`)?

iwanowww replied:
And it would seamlessly cover all possible cases, like static final fields (which can be concurrently updated through Reflection API or Unsafe).

TobiHartmann replied:
But wouldn't that require duplication of the `ciEnv::check_stable_value` logic and having a `GrowableArray` per `ciObject` instance? 

Regarding covering more cases like static final fields: I intentionally omitted them because I don't think it's possible to delay constant folding of such loads until CCP (since `null` is a valid value) and the lookup is done only once during (I)GVN. The regression test includes a corresponding case (see `testFinal`). Even with unsafe accesses, if the field holder is not known at parse time, the access is conservatively marked as mismatched and thus no constant folding is attempted later.

iwanowww replied:
> But wouldn't that require duplication of the ciEnv::check_stable_value logic and having a GrowableArray per ciObject instance?

Are you primarily concerned about footprint? The field can be lazily initialized to reduce possible effects.
`check_stable_value` still can be shared (e.g., placed on `ciObject`).

> Regarding covering more cases like static final fields: I intentionally omitted them because I don't think it's possible to delay constant folding of such loads until CCP

If you want to limit the current patch to stable case only, that's fine with me. But keep in mind that we still want to uniformly treat all cases where field/array element values are considered constant. Concurrent field modifications introduce non-determinism into compilation process which we would like to avoid. So, even if we don't cover final fields right away, let's prepare for a future enhancement which will cover that. In that respect, stable references in names are misleading.

TobiHartmann replied:
My concern is that putting the logic in `ciObject` adds a new field to all subclass instances while it's only rarely needed. Given that a large number of these objects are created during compilation, that adds to footprint even if the cache is initialized lazily.

I still gave this a try and another issue is that using `ciConstant` in `ciObject` creates a circular dependency (`invalid use of incomplete type`) because `ciConstant` depends on `ciObject`. We could of course duplicate the logic and put things both into `ciInstance` and `ciArray` but then a side table in `ciEnv` seems preferable to me.

Until proven otherwise, I would like to limit the current patch to stable cases because I don't think there are any issues for non-stable cases. I can rename the code from "stable" to "constant" but I'm not sure how much sense that makes given it's all limited to and guarded by `FoldStableValues`. What do you think?

iwanowww replied:
Personally, I'm not concerned about footprint increase from a field on `ciObject` class. Those represent oop constants observed by JITs and their count should be modest.

Speaking of relation between `ciConstant` and `ciObject`, `ciConstant.hpp` only refers to `ciObject*`, so the cycle can be broken by prepending `class ciObject;` in `ciConstant.hpp`. 

>  I don't think there are any issues for non-stable cases.
I'm fine with addressing only stable cases with the current patch for now, but I don't agree there's nothing to improve for non-stable cases.

Concurrent field modifications may seem benign, but our previous experience shows that it's safer to work with a consistent view through CI (when it is feasible/possible). 
(There's a relevant RFE filed: https://bugs.openjdk.org/browse/JDK-8294616) And it helps improve compilation replay along the way. And it's not specific to C2.

So, IMO in the longer term CI is the right place to fix it (irrespective whether stable or all constant cases are handled). It's a bit strange to see the new checks added in `opto/type.cpp` (irrespective of where the cache is located).

TobiHartmann replied:
> Speaking of relation between ciConstant and ciObject, ciConstant.hpp only refers to ciObject*, so the cycle can be broken by prepending class ciObject; in ciConstant.hpp.

That won't help. All `ci*` classes, including `ciObject`, are already forward declared in `ciClassList.hpp` which is included in `ciConstant.hpp`. The following usages in `ciConstant` require a complete `ciObject` type:
https://github.com/openjdk/jdk/blob/c5c8c0644d9442846de15422285fffeb91c3e0a1/src/hotspot/share/ci/ciConstant.hpp#L115
https://github.com/openjdk/jdk/blob/c5c8c0644d9442846de15422285fffeb91c3e0a1/src/hotspot/share/ci/ciConstant.hpp#L134

Another issue is that `ciConstant.hpp` includes `ciNullObject.hpp` which inherits from `ciObject`. I worked around both issues by moving code and includes around.

I moved the code to `ciObject`, as you proposed, and extended it to include final fields as well. Let me know what you think.

Testing now revealed a new "Not monotonic" assert that seems to be triggered by caching non-stable final fields. I'll investigate but I'm not able to reproduce it yet.

jatin-bhateja commented:
> Vladimirs, Jatin, thanks for the reviews and discussion!
> 
> The question is what we want to achieve with this patch:
> 
> 1. Prevent C2 from crashing / asserting when observing inconsistent field values
> 2. Prevent C2 from applying optimizations based on inconsistent field values
> 3. Prevent C2 from constant folding stale values
> 
> I think my proposed patch achieves 1) which is also the only issue we ever observed in real code: Racy initialization of method handles for indy string concat in core libraries code. The verification code proposed by ([JDK-8024042](https://bugs.openjdk.org/browse/JDK-8024042)) would catch this. Constant folding a stale value does not matter here, semantically, we can use any of the equivalent method handles that we observe. It just confuses CCP verification. Convergence of CCP is also not an issue because we don't update a node's type indefinitely but only if the type of a relevant input node changed.
> 
> Regarding 2): Couldn't any such wrong/undesired behaviour happen with execution in the interpreter as well? The only difference would be that optimized C2 compiled code would **always** behave that way. But according to the specification for `@Stable`, that's okay:
> 
> https://github.com/openjdk/jdk/blob/5c8c67c523279de728248f54382c40fb20d0ab63/src/java.base/share/classes/jdk/internal/vm/annotation/Stable.java#L74-L80
> 
> Also, doesn't the above specification mean that all non-synchronized initializations of stable fields can lead to undefined behavior?
> 
> > Imaging you read same stable field on separate paths and then compare constants.
> 
> That will also happen with interpreted code when the stable field is written between the two reads.
> 
> The only complete solution would be 3), which, as Vladimir I. already pointed out, does not seem feasible given that we would need to intercept all writes to stable fields and array elements.
> 
> I'm not against caching but I'm wondering how much sense it makes to apply an expensive and complex (partial) fix to C2 while C1 and the interpreter are still affected. Shouldn't the remaining issues be fixed in Java code if undefined/unexpected behavior is ever observed?

Thanks @TobiHartmann for explanations. Caching looks correct. First intercept of constant field folding is dring parsing when a field access is made and later on gvn folds the constant field loads. Similar change could be done for C1 https://github.com/openjdk/jdk/blob/af8d3fb21ab59104d49bd664f634399fb72ecbd2/src/hotspot/share/c1/c1_GraphBuilder.cpp#L1797
But given that problem is only seen during aggressive optimization like CCP I agree with @iwanowww's suggestion of restricting the scope of fix to C2.

TobiHartmann commented:
Thanks for the review, Jatin!

iwanowww commented on src/hotspot/share/ci/ciArray.cpp:
```diff
@@ -95,7 +95,8 @@ ciConstant ciArray::element_value_impl(BasicType elembt,
   GUARDED_VM_ENTRY(
-    return element_value_impl(elembt, get_arrayOop(), index);
+    ciConstant value = element_value_impl(elembt, get_arrayOop(), index);
+    return check_constant_value_cache(index, value);
```
If you check the cache first, you don't need to be in VM when doing that. As 

(As it is now, it doesn't help static/instance fields, but it can be improved. Something for a follow-up cleanup.)

iwanowww commented on src/hotspot/share/ci/ciObject.hpp:
```diff
@@ -57,6 +59,24 @@ class ciObject : public ciBaseObject {
+  // Cache constant value lookups to ensure that consistent values are observed during compilation.
+  class ConstantValue {
+    private:
+      const ciObject* _obj;
```
Since `_constant_values` is instance-specific, there's no need in `_obj` anymore, right?

TobiHartmann replied:
Good catch, removed.

iwanowww reviewed (commented):
Thanks for doing the cleanups, Tobias! I like how the patch looks now.

TobiHartmann commented:
Thanks, Vladimir! I updated the code according to your suggestions.

> As it is now, it doesn't help static/instance fields, but it can be improved. Something for a follow-up cleanup.

Could you elaborate? Final static/instance fields are handled as well (`ciField::constant_value_of` -> `ciInstance::field_value` -> `ciInstance::field_value_impl`). Do you mean non-final fields?

> Testing now revealed a new "Not monotonic" assert that seems to be triggered by caching non-stable final fields. I'll investigate but I'm not able to reproduce it yet.

I reproduced it. The problem is the new call to `verify_type` that I added to `PhaseCCP::verify_analyze`. It's executed before the `LoadNode` bailout:

```
        // MemNode::can_see_stored_value looks up through many memory nodes,
        // which means we would need to notify modifications from far up in
        // the inputs all the way down to the LoadNode. We don't do that.
```

I moved it below the bailout. I think we can also strengthen the verification by only bailing out if the current type is not a singleton, because even for LoadNodes, it should never happen that a node with a singleton type is updated to a different type.

iwanowww commented on src/hotspot/share/ci/ciArray.cpp:
```diff
@@ -94,9 +94,11 @@ ciConstant ciArray::element_value_impl(BasicType elembt,
-    return element_value_impl(elembt, get_arrayOop(), index);
+    value = element_value_impl(elembt, get_arrayOop(), index);
   )
+  return check_constant_value_cache(index, value);
```
It still unconditionally goes through VM transition on every call. If you check the cache first and call `element_value_impl()` only when there's a miss, then you can save on state transition when the cache is populated. But you would need additional call to update the cache.

iwanowww replied:
Or you could move both checks into `element_value_impl()` along with `GUARDED_VM_ENTRY` (and convert it into `VM_ENTRY_MARK`).

iwanowww commented on src/hotspot/share/ci/ciObject.cpp:
```diff
@@ -168,6 +168,26 @@ jobject ciObject::constant_encoding() {
+  }
+  for (int i = 0; i < _constant_values->length(); ++i) {
+    ConstantValue cached_val = _constant_values->at(i);
+    if (cached_val.off() == off) {
```
I'd also check that basic types matches (either with a runtime check or an assert).

TobiHartmann replied:
I added an assert that immediately triggered because `ciArray::element_value_impl` creates a `ciConstant` with basic type `T_OBJECT` for a `T_ARRAY` array element. I fixed that code.

iwanowww commented:
>> As it is now, it doesn't help static/instance fields, but it can be improved. Something for a follow-up cleanup.
> Could you elaborate? Final static/instance fields are handled as well (ciField::constant_value_of -> ciInstance::field_value -> ciInstance::field_value_impl). Do you mean non-final fields?

Not necessarily. I spotted the following for static final fields:
```
ciConstant ciField::constant_value() {
  ...
  if (_constant_value.basic_type() == T_ILLEGAL) {
    // Static fields are placed in mirror objects.
    VM_ENTRY_MARK;
    ciInstance* mirror = CURRENT_ENV->get_instance(_holder->get_Klass()->java_mirror());
    _constant_value = mirror->field_value_impl(type()->basic_type(), offset());
  }
```
As it is now, you can't avoid the state transition when doing `field_value_impl()` call. 
You could reshape it into:
```
  if (_constant_value.basic_type() == T_ILLEGAL) {
    // Static fields are placed in mirror objects.
    ciInstance* mirror = _holder->java_mirror();
    _constant_value = mirror->field_value_impl(type()->basic_type(), offset());
  }
``` 
... and then put a call to check the cache first before trying to load a value from the field. There should be a `VM_ENTRY_MARK` in `field_value_impl()`, but the check can be placed before it. Then `ciInstance::field_value()` won't need `GUARDED_VM_ENTRY` when calling into `field_value_impl()`.

iwanowww commented:
> I reproduced it. The problem is the new call to verify_type that I added to PhaseCCP::verify_analyze.

Nice catch!

TobiHartmann commented:
Thanks again, Vladimir. I updated the change according to your suggestions.

iwanowww commented on src/hotspot/share/ci/ciObject.cpp:
```diff
@@ -168,6 +168,37 @@ jobject ciObject::constant_encoding() {
+  assert(val.is_valid(), "value must be valid");
+  if (_constant_values == nullptr) {
+    Arena* arena = CURRENT_ENV->arena();
+    _constant_values = new (arena) GrowableArray<ConstantValue>(arena, 0, 0, ConstantValue());
```
Considering an element is added right away, it makes sense to increase initial capacity (to 1, at least).

TobiHartmann replied:
Makes sense. Done.

iwanowww commented on src/hotspot/share/ci/ciObject.cpp:
```diff
@@ -168,6 +168,37 @@ jobject ciObject::constant_encoding() {
+//
+// Add a constant value to the cache.
+void ciObject::add_to_constant_value_cache(int off, ciConstant val) {
+  assert(val.is_valid(), "value must be valid");
```
Maybe add an assert to ensure the list doesn't contain the value?
```
assert(check_constant_value_cache(off, val.basic_type()).is_valid() == false, "duplicate");
```

TobiHartmann replied:
Done.

iwanowww reviewed (approved):
Overall, looks good!

TobiHartmann commented:
Thanks again, Vladimir. I incorporated both your suggestions.

iwanowww reviewed (approved):
Looks good.

TobiHartmann commented:
Thanks Vladimir. @vnkozlov, @chhagedorn, @jatin-bhateja are you okay with the latest version as well?

vnkozlov reviewed (commented):
I am fine with moving caching to CI.

As I see (may be wrong) you allow C2 to see all cached values which may introducing inconsistency in generated code. I am not sure how `PhaseCPP::verify_type()` works in such case.

TobiHartmann commented on src/hotspot/share/ci/ciObject.hpp:
```diff
@@ -57,6 +59,22 @@ class ciObject : public ciBaseObject {
+
+    public:
+      ConstantValue() : _off(0), _value(ciConstant()) { }
+      ConstantValue(int off, ciConstant value) : _off(off), _value(value) { }
```
No particular reason. I switched the order of the fields.

TobiHartmann commented:
> As I see (may be wrong) you allow C2 to see all cached values which may introducing inconsistency in generated code. I am not sure how PhaseCPP::verify_type() works in such case.

I'm not sure if I understand your concern, what do you mean by "all cached values"? The cache will only contain values observed during the current compilation (it's lifetime is limited by the lifetime of `ciObject` which is compile task specific) and it will ensure that the same value is used for constant folding of a specific field or array element.

vnkozlov commented:
> > As I see (may be wrong) you allow C2 to see all cached values which may introducing inconsistency in generated code. I am not sure how PhaseCPP::verify_type() works in such case.
> 
> I'm not sure if I understand your concern, what do you mean by "all cached values"? The cache will only contain values observed during the current compilation (it's lifetime is limited by the lifetime of `ciObject` which is compile task specific) and it will ensure that the same value is used for constant folding of a specific field or array element.

After reading Vladimir's and your discussion and looking on code I see that you check value in cache before adding it to cache. So my concern is mute. But why do you need `GrowableArray` in such case? That confused me because I first thought you cache all values.

TobiHartmann commented:
> But why do you need GrowableArray in such case? That confused me because I first thought you cache all values.

Because there's only a very limited number of constant fields per compilation, so I thought a simple array is good enough. Do you prefer a hash map (`Dict` or something) instead?

vnkozlov commented:
> > But why do you need GrowableArray in such case? That confused me because I first thought you cache all values.
> 
> Because there's only a very limited number of constant fields per compilation, so I thought a simple array is good enough. Do you prefer a hash map (`Dict` or something) instead?

My mistake. I thought you collected values per field. But you collected all constant fields per object. Good. GrowableArray is perfectly fine for that. Finally I got this ;^)

One last question. Do I understand correctly that static fields constant value is already handled by `ciField::constant_value()`?

TobiHartmann commented:
Thanks again, Vladimir.

> One last question. Do I understand correctly that static fields constant value is already handled by ciField::constant_value()?

Yes, for example for C2, final static fields are handled via `Type::make_constant_from_field` -> `ciField::constant_value` -> `ciInstance::field_value_impl` (on the mirror object) which will then check the cache via `check_constant_value_cache`.

vnkozlov reviewed (approved):
Good. Thank you for answering all my questions.
